### PR TITLE
[`iter_without_into_iter`]: fix papercuts in suggestion and restrict linting to exported types

### PR DESCRIPTION
**Pull request by y21**
See #11692 for more context.

tldr: the lint `iter_without_into_iter` has suggestions that don't compile, which imo isn't that problematic because it does have the appropriate `Applicability` that tells external tools that it shouldn't be auto-applied.
However there were some obvious "errors" in the suggestion that really should've been included in my initial PR adding the lint, which is fixed by this PR:
- `IntoIterator::into_iter` needs a `self` argument.
- `IntoIterator::Iter` associated type doesn't exist. This should've just been `Item`.

This still doesn't make it machine applicable, and the remaining things are imho quite non-trivial to implement, as I've explained in https://github.com/rust-lang/rust-clippy/issues/11692#issuecomment-1773886111.
I personally think it's fine to leave it there and let the user change the remaining errors when copy-pasting the suggestion (e.g. errors caused by lifetimes that were permitted in fn return-position but are not in associated types).
This is how many of our other lint suggestions already work.

Also, we now restrict linting to only exported types. This required moving basically all of the tests around since they were previously in the `main` function. Same for `into_iter_without_iter`. The git diff is a bit useless here...

changelog: [`iter_without_into_iter`]: fix papercuts in suggestion and restrict linting to exported types

(cc @lopopolo, figured I should mention you since you created the issue)

### CODE REVIEW

File: clippy_lints/src/iter_without_into_iter.rs [hunk 1]
```diff
@@ -104,6 +104,12 @@ fn is_nameable_in_impl_trait(ty: &rustc_hir::Ty<'_>) -> bool {
     !matches!(ty.kind, TyKind::OpaqueDef(..))
 }
 
+fn is_ty_exported(cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
```
Comment (xFrednet) [inline clippy_lints/src/iter_without_into_iter.rs:107]: I think it would be good to document this restriction to public items in the lint description. Maybe under a `### Limitations` section to just under the normal `### What it does?` section.

**Review by xFrednet — APPROVED** [review]
LGTM, I have one small suggestion, and then you can `r=me`.

Clippy's documentation says that this lint will be stabilized in rust 1.74. Do you think this is worth a beta backport? It's a pedantic lint, which doesn't make it super necessary, but the restriction to only exported items, might make a big difference, to the amount of lint triggers :thinking:

File: clippy_lints/src/iter_without_into_iter.rs (hunk 1 above)
Comment (y21) [reply clippy_lints/src/iter_without_into_iter.rs:107]: Good idea. In my experience, "Limitations" is under "Why is this bad?". I'll put that there instead, for consistency's sake.

Comment (y21) [reply clippy_lints/src/iter_without_into_iter.rs:118]: Done. What do you think about the wording? Not gonna approve just yet because I'd like to hear your opinion

**Comment by y21** [comment]
> Do you think this is worth a beta backport?

I think it's a good idea, even for a pedantic lint. It's not all that useful for private types imo and I think it could make a noticeable difference in how often the lint occurs. The errors in the suggestion were also pretty obvious so it would be nice if this fix gets into beta along with the PR that introduced this.
But if it's too much of a hassle or this is the only PR that is gonna be nominated for backporting, it might be unnecessary.

**Comment by lopopolo** [comment]
Please backport this. Paper cuts in clippy lints lead to them being disabled in my codebases. I almost never revisit an `!#[allow(clippy::xxx)]` once it gets added

**Comment by xFrednet** [comment]
I agree and this shouldn't be a problem, that's why I asked. `@flip1995` Usually does the beta back ports, a week or so before the release. To indicate that a fix should be back ported, we just need to add the `beta-nominated` label.

@rustbot label +beta-nominated

File: clippy_lints/src/iter_without_into_iter.rs (hunk 1 above)
Comment (y21) [reply clippy_lints/src/iter_without_into_iter.rs:118]: @xFrednet ? :)
(I'm not sure if you saw the comment or if you're just busy)

Comment (xFrednet) [reply clippy_lints/src/iter_without_into_iter.rs:118]: I actually missed your comment, thank you for pointing that out. It's correct, that the `### Limitations` section comes after `### Why is it bad?`

File: clippy_lints/src/iter_without_into_iter.rs [hunk 2]
```diff
@@ -61,6 +66,12 @@ declare_clippy_lint! {
     /// by just calling `.iter()`, instead of the more awkward `<&Type>::into_iter` or `(&val).into_iter()` syntax
     /// in case of ambiguity with another `IntoIterator` impl.
     ///
+    /// ### Limitations
+    /// This lint is restricted to exported types only, because it is aimed at guiding towards an
+    /// idiomatic, _public_ API.
+    /// Adding an `iter` or `iter_mut` for private types when it is not needed or used doesn't improve code,
+    /// and in fact, is linted against by the `dead_code` lint.
+    ///
```
Comment (xFrednet) [inline clippy_lints/src/iter_without_into_iter.rs:74]: The current version is fine. I would formulate it a bit differently, but this is mostly a question of taste. I leave the decision up to you, which one you like more, or if you want to combine the two versions :)

```
    /// ### Limitations
    /// This lint focusess on providing an idiomatic API. Therefore, it will only
    /// lint on types, which are accessible outside of the crate. For internal types
    /// these methods can be added on demand, if they are acually needed. Otherwise,
    /// it would trigger the [`dead_code`] lint for the unused method.
    ///
    /// [`dead_code`]: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#dead-code
```

Comment (y21) [reply clippy_lints/src/iter_without_into_iter.rs]: Adjusted the documentation a bit with your suggestions in mind.

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
LGTM, thank you for the update :)

@bors r+

**Comment by flip1995** [comment]
This lint does not exist in `beta` yet and this PR was already synced to rustc. So this PR will be included when this lint gets to `beta` and ultimately `stable`. => Nothing to do for me. My favorite kind of backports :upside_down_face:
